handler/src/main/java/io/netty/handler/ssl/ReferenceCountedOpenSslEngine.java

```diff
@@ -622,52 +614,50 @@ public final SSLEngineResult wrap(
                     if (result > 0) {
                         bytesConsumed += result;
-
-                        pendingNetResult = readPendingBytesFromBIO(dst, bytesConsumed, bytesProduced, status);
-                        if (pendingNetResult != null) {
-                            if (pendingNetResult.getStatus() != OK) {
-                                return pendingNetResult;
-                            }
-                            bytesProduced = pendingNetResult.bytesProduced();
-                        }
-                        if (bytesConsumed == MAX_PLAINTEXT_LENGTH) {
-                            // If we consumed the maximum amount of bytes for the plaintext length break out of the
-                            // loop and start to fill the dst buffer.
-                            break loop;
-                        }
                     } else {
                         int sslError = SSL.getError(ssl, result);
                         switch (sslError) {
                             case SSL.SSL_ERROR_ZERO_RETURN:
                                 // This means the connection was shutdown correctly, close inbound and outbound
                                 if (!receivedShutdown) {
                                     closeAll();
+
+                                    // Check to see if the engine wrote data into the network BIO as closeAll() may
+                                    // produce an alert.
+                                    pendingNet = SSL.pendingWrittenBytesInBIO(networkBIO);
                                 }
-                                pendingNetResult = readPendingBytesFromBIO(dst, bytesConsumed, bytesProduced, status);
-                                return pendingNetResult != null ? pendingNetResult : CLOSED_NOT_HANDSHAKING;
+
+                                return pendingNet > 0 ?
+                                        readPendingBytesFromBIO(dst, bytesConsumed, bytesProduced, pendingNet, status) :
+                                        new SSLEngineResult(CLOSED, NOT_HANDSHAKING, bytesConsumed, bytesProduced);
+
                             case SSL.SSL_ERROR_WANT_READ:
                                 // If there is no pending data to read from BIO we should go back to event loop and try
                                 // to read more data [1]. It is also possible that event loop will detect the socket
                                 // has been closed. [1] https://www.openssl.org/docs/manmaster/ssl/SSL_write.html
-                                pendingNetResult = readPendingBytesFromBIO(dst, bytesConsumed, bytesProduced, status);
-                                return pendingNetResult != null ? pendingNetResult :
-                                        new SSLEngineResult(isOutboundDone() ? CLOSED : OK,
-                                                NEED_UNWRAP, bytesConsumed, bytesProduced);
+
+                                if (pendingNet <= 0) {
```

Comment:
in this case src could potentially have more bytes to SSL_write. This would be a bit pessimistic to return to the loop.